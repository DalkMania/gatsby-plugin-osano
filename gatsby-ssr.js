import React from "react"

// Ensure that the Osano script gets loaded as early as it can. To comply with GDPR Rules
export const onPreRenderHTML = ({ getHeadComponents, replaceHeadComponents }, pluginOptions) => {
  if (
    process.env.NODE_ENV === `production` ||
    pluginOptions.includeInDevelopment
  ) {
    const headComponents = getHeadComponents()
    headComponents.sort((x, y) => {
      if (x.type === 'script' && x.key === 'gatsby-plugin-osano') {
        return -1
      } else if (y.type === 'script' && y.key === 'gatsby-plugin-osano') {
        return 1
      }
      return 0
    })
    replaceHeadComponents(headComponents)
  }
}

export const onRenderBody = ({ setHeadComponents }, pluginOptions) => {
  if (
    process.env.NODE_ENV === `production` ||
    pluginOptions.includeInDevelopment
  ) {
    return setHeadComponents([
      <script
        key={`gatsby-plugin-osano`}
        src={`https://cmp.osano.com/${pluginOptions.customerId}/${pluginOptions.ccid}/osano.js`}
      />,
    ])
  }
}