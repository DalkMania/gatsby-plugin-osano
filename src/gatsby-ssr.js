import React from "react";

// Ensure that the Osano script gets loaded as early as it can. To comply with GDPR Rules
export const onPreRenderHTML = ({ getHeadComponents, replaceHeadComponents }, pluginOptions) => {
    if (process.env.NODE_ENV === `production` || pluginOptions.includeInDevelopment) {
        let headComponents = getHeadComponents();
        let scripts = headComponents.filter((el) => el.type === "script");
        const osano = scripts.find((el) => el.key === "gatsby-plugin-osano");
        scripts = scripts.filter((el) => el !== osano);
        headComponents = headComponents.filter((el) => el.type !== "script");
        scripts.unshift(osano);
        headComponents = headComponents.concat(scripts);

        replaceHeadComponents(headComponents);
    }
};

export const onRenderBody = ({ setHeadComponents }, pluginOptions) => {
    if (process.env.NODE_ENV === `production` || pluginOptions.includeInDevelopment) {
        return setHeadComponents([
            <link
                key={`preload-gatsby-plugin-osano`}
                rel="preload"
                as="script"
                href={`https://cmp.osano.com/${pluginOptions.customerId}/${pluginOptions.ccid}/osano.js`}
            />,
            <script
                key={`gatsby-plugin-osano`}
                src={`https://cmp.osano.com/${pluginOptions.customerId}/${pluginOptions.ccid}/osano.js`}
            />
        ]);
    }
};
