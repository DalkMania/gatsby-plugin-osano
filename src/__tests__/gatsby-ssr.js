import { onPreRenderHTML, onRenderBody } from "../gatsby-ssr";

describe(`gatsby-plugin-osano`, () => {
    describe(`onPreRenderHTML`, () => {
        describe(`in development mode`, () => {
            it(`it does not replace any head components, if the includeInDevelopment option is false`, () => {
                const pluginOptions = { customerId: 1, ccid: 1, includeInDevelopment: false };
                const getHeadComponents = jest.fn();
                const replaceHeadComponents = jest.fn();

                onPreRenderHTML({ getHeadComponents, replaceHeadComponents }, pluginOptions);
                expect(getHeadComponents).not.toHaveBeenCalled();
                expect(replaceHeadComponents).not.toHaveBeenCalled();
            });

            it(`it does replace head components, if the includeInDevelopment option is true and reorders the osano script to be first`, () => {
                const pluginOptions = { customerId: 1, ccid: 1, includeInDevelopment: true };
                const components = [
                    {
                        type: "script",
                        key: `script-1234`
                    },
                    {
                        type: "script",
                        key: `gatsby-plugin-osano`
                    }
                ];
                const sortedComponents = [
                    {
                        type: "script",
                        key: `gatsby-plugin-osano`
                    },
                    {
                        type: "script",
                        key: `script-1234`
                    }
                ];
                const getHeadComponents = jest.fn(() => components);
                const replaceHeadComponents = jest.fn();

                onPreRenderHTML({ getHeadComponents, replaceHeadComponents }, pluginOptions);

                expect(getHeadComponents.mock.calls).toMatchSnapshot();
                expect(getHeadComponents).toHaveBeenCalledTimes(1);
                expect(replaceHeadComponents).toHaveBeenCalledTimes(1);
                expect(replaceHeadComponents).toHaveBeenCalledWith(sortedComponents);
            });
        });

        describe("in production mode", () => {
            let env;

            beforeAll(() => {
                env = process.env.NODE_ENV;
                process.env.NODE_ENV = `production`;
            });

            afterAll(() => {
                process.env.NODE_ENV = env;
            });

            it(`it does replace head components, in production and reorders the osano script to be first`, () => {
                const pluginOptions = { customerId: 1, ccid: 1, includeInDevelopment: true };
                const components = [
                    {
                        type: "script",
                        key: `script-1234`
                    },
                    {
                        type: "script",
                        key: `gatsby-plugin-osano`
                    }
                ];
                const sortedComponents = [
                    {
                        type: "script",
                        key: `gatsby-plugin-osano`
                    },
                    {
                        type: "script",
                        key: `script-1234`
                    }
                ];
                const getHeadComponents = jest.fn(() => components);
                const replaceHeadComponents = jest.fn();

                onPreRenderHTML({ getHeadComponents, replaceHeadComponents }, pluginOptions);

                expect(getHeadComponents.mock.calls).toMatchSnapshot();
                expect(getHeadComponents).toHaveBeenCalledTimes(1);
                expect(replaceHeadComponents).toHaveBeenCalledTimes(1);
                expect(replaceHeadComponents).toHaveBeenCalledWith(sortedComponents);
            });
        });
    });

    describe(`onRenderBody`, () => {
        describe(`in development mode`, () => {
            it(`does not set any head component, if the includeInDevelopment option is false`, () => {
                const setHeadComponents = jest.fn();
                const pluginOptions = { customerId: 1, ccid: 1, includeInDevelopment: false };

                onRenderBody({ setHeadComponents }, pluginOptions);
                expect(setHeadComponents).not.toHaveBeenCalled();
            });

            it(`does set head component, if the includeInDevelopment option is true`, () => {
                const setHeadComponents = jest.fn();
                const pluginOptions = { customerId: 1, ccid: 1, includeInDevelopment: true };

                onRenderBody({ setHeadComponents }, pluginOptions);
                expect(setHeadComponents.mock.calls).toMatchSnapshot();
                expect(setHeadComponents).toHaveBeenCalledTimes(1);
                expect(setHeadComponents).toHaveBeenCalledWith([
                    expect.objectContaining({ key: `preload-gatsby-plugin-osano` }),
                    expect.objectContaining({ key: `gatsby-plugin-osano` })
                ]);
            });
        });

        describe(`in production mode`, () => {
            let env;

            beforeAll(() => {
                env = process.env.NODE_ENV;
                process.env.NODE_ENV = `production`;
            });

            afterAll(() => {
                process.env.NODE_ENV = env;
            });

            it(`does set head component`, () => {
                const setHeadComponents = jest.fn();
                const pluginOptions = { customerId: 1, ccid: 1 };

                onRenderBody({ setHeadComponents }, pluginOptions);
                expect(setHeadComponents.mock.calls).toMatchSnapshot();
                expect(setHeadComponents).toHaveBeenCalledTimes(1);
                expect(setHeadComponents).toHaveBeenCalledWith([
                    expect.objectContaining({ key: `preload-gatsby-plugin-osano` }),
                    expect.objectContaining({ key: `gatsby-plugin-osano` })
                ]);
            });
        });
    });
});
