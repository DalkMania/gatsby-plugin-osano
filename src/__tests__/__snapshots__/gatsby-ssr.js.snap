// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gatsby-plugin-osano onPreRenderHTML in development mode it does replace head components, if the includeInDevelopment option is true and reorders the osano script to be first 1`] = `
Array [
  Array [],
]
`;

exports[`gatsby-plugin-osano onPreRenderHTML in production mode it does replace head components, in production and reorders the osano script to be first 1`] = `
Array [
  Array [],
]
`;

exports[`gatsby-plugin-osano onRenderBody in development mode does set head component, if the includeInDevelopment option is true 1`] = `
Array [
  Array [
    Array [
      <script
        src="https://cmp.osano.com/1/1/osano.js"
      />,
    ],
  ],
]
`;

exports[`gatsby-plugin-osano onRenderBody in production mode does set head component 1`] = `
Array [
  Array [
    Array [
      <script
        src="https://cmp.osano.com/1/1/osano.js"
      />,
    ],
  ],
]
`;
